src/metrics.jl

```diff
@@ -99,8 +99,16 @@ struct MeanSqDeviation <: SemiMetric end
 struct RMSDeviation <: Metric end
 struct NormRMSDeviation <: Metric end
 
+struct PeriodicEuclidean{N,T <: Real} <: Metric
+    periods::NTuple{N,T}
+end
+PeriodicEuclidean(periods::Tuple) = PeriodicEuclidean(promote(periods...))
+PeriodicEuclidean(periods::AbstractVector{T}) where {T<:Real} = PeriodicEuclidean{length(periods),T}(tuple(periods...))
```

Comment:
Why not require a tuple input?

Reply:
The idea here was to be liberal at construction, and make sure a tuple is constructed internally. Also, I had in mind adopting this approach to the weighted metrics. There, you can pass vectors of weights, and adopting the approach of this PR (with insisting on tuples) would be breaking for the weighted metrics.

Reply:
AFAICT the situation is quite different for periods and for weights: the number of weights varies with the number of observations. OTC, the number of periods isn't supposed to vary much I guess, since it is a type parameter. But maybe that parameter should be dropped and periods should be stored as a vector instead? That would be consistent with the decision not to merge https://github.com/JuliaStats/Distances.jl/pull/115.

Reply:
This is exactly how I started out. Copying the weighted metric code, making the minimal modifications necessary for the periodic metric, and let go. That's when @KristofferC pointed out that code duplication is not good, and was pointing to an already existing issue with it (optimization of the for-loop in `evaluate`). But in fact, that same "criticism" could have been made earlier for the weighted metrics. Note that (regardless of the implementation) the weighted metrics and the periodic metric are of the exact same type: you have one parameter per observation. So @KristofferC 's suggestion was to somehow "hide" the dimensionwise parameter in an additional state in the `s` variable, and otherwise use the standard `evaluate` instance. The current state of this PR contains my take at this approach, and I do think that tuples are the right thing to use for type inference via `Base.tail` etc. The "fixed length" tuple thing is just due to my ignorance of a better type for this, but I just noticed that there is something like `Vararg{T}`, and I'll see if I can get this running.

Reply:
OK, if the length of the periods is the same as the number of observations, then making it a type parameter is probably not appropriate. Better use a vector than a `Vararg`, though.

Reply:
But then the whole evaluate construction breaks down. It would force me to use indexing instead, which allocates...?

Reply:
Can't you just pass the index where to start instead of calling `Base.tail(s1[2])` or its vector equivalent `s1[2][2:end]`?

Reply:
I'm afraid I can't pass an index because there is no such interface in the `UnionMetrics` `evaluate`, which I'm supposed (and want) to use. The vector equivalent allocates and doubles the runtime in the small example in the docstring, in comparison to my `NTuple` version.

Generally, what's the issue with having an `NTuple` for the parameters?

Reply:
Nevermind, the example was too specific. For the `NTuple` version, there was a significant slow-down for dimension 3 and higher. The slowdown is by far not as dramatic for `Vector`s. Unfortunately, for `Vector`s there are as many allocations as dimensions. I'll investigate further.

Reply:
Assuming that we're aiming at bringing all `UnionMetrics` under one `evaluate` umbrella, we could change the signature for `eval_op` throughout as follows:
```
eval_op(::Metric, ai, bi, i::Int=1)
```
As for `PeriodicEuclidean`, this yields the fastest implementation I could ever achieve. One would have to change all other `eval_op`s accordingly, most of which would not use it. This would also allow to "merge" the weighted metrics with the unweighted ones.

Reply:
And what happens if you use `Iterators.drop(s1[2], 1)`?

Reply:
That seems to allocate and yield slower computations. I kind of like the above `eval_op` idea, since it mimicks what we have for the weighted metrics, it would not be breaking existing code, and it shouldn't slow down the classic metrics, should it?

Reply:
Sounds like a good solution (though I don't master the whole machinery of this package).

Reply:
Unfortunately, this additional argument with default value seems to affect the classical metrics performancewise, which would not use it. Unless there are other ideas, I would suggest to merge the weighted metrics and the periodic metric here in one file, and declare them as (something like) `ParametricMetrics`. These would have their own `evaluate` instance and an additional parameter `i::Int=1` in their respective `eval_op`. This would also form a generic framework for metrics that have dimensionwise parameters. I'll do the changes and ask for comments.

Reply:
How can it have an effect on performance if `i` isn't used? :-/

Reply:
I think it is because when `eval_op` is called with only 3 arguments `(::Metric, a, b)`, then it calls internally the 4-args instance and appends the default value `1`, which needs to be created. By "not used" I mean that it's value is not used in the function body, but the call routines still change.

Reply:
All of this should happen at compile time and be optimized out by the compiler (at least with `@inline`). Can you show the code?